### ISSUE TITLE
Creating table from list of rows fails with numpy-dev (OK for 1.17)

### ISSUE THREAD

**Issue by duncanmmacleod** [astropy/astropy#8976]
With `astropy-3.2.1` and `numpy-1.16.4` I can create one table from the rows of another, as follows:

```pycon
>>> from astropy.table import Table
>>> t = Table([[1, 2, 3], [4, 5, 6]], names=('a', 'b'))
>>> print(t)
 a   b
--- ---
  1   4
  2   5
  3   6
>>> print(type(t)(rows=[row for row in t], names=t.dtype.names))
 a   b
--- ---
  1   4
  2   5
  3   6
```

However, with the same `astropy`, and `numpy-1.17.0rc1` I cannot (well, I can, it just doesn't do what I expect it to):

```pycon
>>> from astropy.table import Table
>>> t = Table([[1, 2, 3], [4, 5, 6]], names=('a', 'b'))
>>> print(t)
 a   b
--- ---
  1   4
  2   5
  3   6
>>> print(type(t)(rows=[row for row in t], names=t.dtype.names))
a [2]  b [2]
------ ------
1 .. 4 1 .. 4
2 .. 5 2 .. 5
3 .. 6 3 .. 6
```

If I manually cast each row to list, then things do work again:

```pycon
>>> print(type(t)(rows=[list(row) for row in t], names=t.dtype.names))
 a   b
--- ---
  1   4
  2   5
  3   6
```

Is this just a consequence of an upstream change in the new pre-release of `numpy`, or are tables not really supposed to be created this way in the first place?

**Comment by bsipocz** [comment]
I see the same with master, too.

ping @mhvk

**Comment by mhvk** [comment]
Definitely a bug. Inspecting code a bit more, the piece that changed behaviour is an [`np.array()` in astropy.table.np_utils.recarray_fromrecords](https://github.com/astropy/astropy/blob/master/astropy/table/np_utils.py#L190). Specifically,
```
rec_list = [row for row in t]
np.array(rec_list, dtype=object)
# numpy 1.16
array([[1, 4],
       [2, 5],
       [3, 6]], dtype=object)
# numpy 1.17rc
array([[(1, 4), (1, 4)],
       [(2, 5), (2, 5)],
       [(3, 6), (3, 6)]], dtype=object)
# in both cases
type(obj[0])
# numpy.ndarray
```

I briefly tried to make a minimal test case for a possible issue with numpy, but it ended up not that minimal: https://github.com/numpy/numpy/issues/13958: the `np.array` call on `Row` seems not just to take its `__array__` method, but also look at its length.

Overall, it may be good to rewrite `recarray_fromrecords` - possibly just special-casing `Row`.

**Comment by mhvk** [comment]
Note that while the problem is caused upstream, I think it might still be good to fix `recarray_fromrecords` regardless, at it seems more than a little flaky.

**Comment by mhvk** [comment]
Numpy has reverted the change so things should work again (for 1.17 at least). If someone is willing to make a test case out of this, that would be great!

**Comment by pllim** [comment]
Since the test case is already presented in the original post, it is just a matter of copy-paste? I'll take a stab at this while I procrastinate from working on my talk.

**Comment by mhvk** [comment]
Reopening since while things will now work in numpy 1.17, the revert was not done on `numpy-dev`; my sense is that we may just want to avoid the problem altogether by fixing `recarray_fromrecords`.

**Comment by pllim** [comment]
I am afraid I don't have enough expertise on how to special-casing `Row` properly here. 😢 

https://github.com/astropy/astropy/blob/0c4ea46e5bda8405f4a6ed8da9bdd2738a18385b/astropy/table/np_utils.py#L176-L196

**Comment by mhvk** [comment]
Just noticed that this will get fixed by #9165.
